### PR TITLE
Allow reserved words in unquoted imports but disallow whitespace.

### LINKED ISSUE #3983: URI shorthands without internal whitespace.
The [proposal for URI shorthands](https://github.com/dart-lang/language/blob/main/working/unquoted-imports/feature-specification.md) allows sequence of tokens of the form: `<dottedIdentifierList> ('/' <dottedIdentifierList>)*` as unquoted imports.

It's a syntactic grammar, not a lexical grammar, which means that it does not affect tokenization, which would otherwise need to be context sensitive, and so it can reuse existing grammar productions.

As currently specified, it allows whitespace and comments between the tokens (it doesn't say it doesn't, and syntactic grammar rules do allow whitespace).

I propose to *not* allow whitespace or comments inside the URI shorthand.
Further, I suggest to *require* the URI shorthand to be terminated by a whitespace or semicolon. Any other adjacent token will *look* like it should be part of the URI, but it isn't. It's better to make the URI clearly delimited, by whitespace (with no internal whitespace) or with a final `;` which users are used to reading at end of lines.

This change would not affect tokenization, it would just stop parsing the token sequence as an unquoted URI when there is whitespace or comments after a token, and it would make it an error to have any non-allowed token touching the URI.

I believe this to be *less error-prone*, especially around `show`/`hide` and that any actual internal whitespace or comments will hurt readability signficantly.
If someone takes a quoted URI and removes just the quotes, `package:` and `.dart`, they may get any number of parsing errors if the URI is not of a form supported by the shorthand. By including all the tokens up to the first whitespace, it's easier to point to the position *inside* the supposed URI where the unsupported character is, rather than reporting those tokens as "unexpected" after the URI. _Including the tokens in the URI is, my hypothesis, more likely to match user intent, and therefore doing so gives more useful error messages._

Not allowing other following tokens should extend to comments. This would even disallow adjacent EOL-comments, so `import foo/bar//baz;` would be an *error*, because it's misleading. If the import should end there, adding a space is not a problem. If it shouldn't, it's probably an error.
(This does affect formatting. If someone writes:
```dart
import foo/bar// come comment here
  as banana;
```
and expects the formatter to insert a space, it won't. That is a real issue, and may be enough reason to allow an EOL-comment to follow an unquoted URI anyway. Even if it then won't help with a `import foo/bar//baz;` mistyping.)

This also includes *string literals* in the included tokens, which is highly confusing.
It should be an immediate error if a string literal is adjacent to an unquoted URI.
Since a string literal may contain whitespace, it feels weird to continue the unquoted URI after the string.
We may want to consider making string literals terminate unquoted URIs too, and then it's always an error if they're following the unquoted URI, adjacent or not. (In that case, an unquoted URI is terminated by a whitespace, semicolon or string literal).

While the grammar is unambiguous (because there must be a `.` or `/` between any two identifiers), it can require some look-ahead to figure out what an *error* should be. That is, it may make error recovery harder.

Example:
```dart
import mypkg/src/show. as hide hide show show banana;
```
This is invalid because it parses as `.../show.as hide hide show show banana;` and there is a comma missing before `banana`.
If the `.` was the mistake, then removing it would make the import valid as `.../show as hide hide show show banana;` with all of an `as`, `hide` and `show` clause. (The `as` identifier is a built-in identifier, `hide` and `show` are only contextually keywords.)
Nobody writes both `hide` and `show`. _Could we make unquoted imports only allow one of `show` and `hide`, just to make things easier for ourselves? You never *need* more than one._

Some error reporting examples:

```dart
import mypkg/src/time. show banana; // Forgot to delete the `.` when deleting `.dart`.
import mypkg/src/time show banana; // What was intended.
```

If internal whitespace is not allowed, the first import would give an error of "missing identifier after `.`" rather than "unexpected token: banana".

```dart
import mypkg/src//time;
import show;
```
Here the error will be "unexpected token: `import`" unless we disallow adjacent EOL-comments, then it will be "`//time;` comment not allowed in unquoted URI.

```dart
import mypkg/src/time+date;
```
would give an error of "`+` not allowed in unquoted URI" rather than "unexpected identifier `+`, expected on of `as`, `show`, `hide` or `;`".

This would have the effect of making whitespace significant, and therefore it can affect automatic formatting.
That is another reason I want to include all adjacent tokens into the URI. The code can still be formatted, with its malformed URI.

Grammatically, the grammar of the unquoted URI would be:
```ebnf
-- Grammar rules
<uri> ::= <stringLiteral> | <unquotedUri> 
<unquotedUri> ::= <unquotedUriToken> (~(WS|';') <unquotedUriToken>)* LOOKAHEAD(WHITESPACE | ';')
<validUnquotedUri> ::= <uriWords> (~WS '/' ~WS <uriWords>)*
<uriWords> ::= (<uriWord> ~WS '.' ~WS)* <uriWord>
<uriWord> ::= <identifier>

<unquotedUriToken> ::= *any* token other than WHITESPACE, COMMENT or ';'.

-- Tokenizer rule
WS := WHITESPACE | COMMENT
```
where the `~WS` is a one-token negative lookahead, and `LOOKAHEAD(...)` is a positive look-ahead.
The parsing will collect any token up to the next whitespace or semicolon, then check if those tokens can be produced by `<validUnquotedUri>`, rather than trying to only collect enough to be valid, and then leave the rest.

This also has the advantage that if we extend the grammar of valid unquoted URIs, we don't need to change the code that collects the URIs, only the validation.

TL;DR: The unquoted URI is undelimited by punctuation. I think it's important for readability and error recovery to make it clearly delimited by *whitespace*, which can be done by not allowing internal whitespace or comments, including tokens until a semicolon or whitespace (or possibly a `/*..*/` comment, but not a `//comment` because it can be a mistyped path), and reporting an error if those tokens do not form a valid unquoted URI, rather than just stopping before the unexpected adjacent token.
I believe this has a higher chance of catching mistakes that assume more characters are allowed in unquoted URIs than actually the case, and therefore giving better error messages. 

In practice, all unquoted URIs will be terminated by whitespace or semicolon, so the restriction only really affects error recovery and error reporting (positively).

### LINKED ISSUE #3984: URI shorthands, allow reserved words.
I propose to allow *reserved words* to occur in unquoted URIs like any other identifier, because URI/file-system paths are not Dart code, and have no reason to be restricted by what happens to be Dart reserved words.

The [proposal for URI shorthands](https://github.com/dart-lang/language/blob/main/working/unquoted-imports/feature-specification.md) allows sequence of tokens of the form: `<dottedIdentifierList> ('/' <dottedIdentifierList>)*` as unquoted imports.

It's a syntactic grammar, not a lexical grammar, which means that it does not affect tokenization, which would otherwise need to be context sensitive, and so it can reuse existing grammar productions.

An example import could be `import somepackage/somelibrary;`.

By using `<dottedIdentifier>` it only allows *identifiers* as parts of the path segments of the URI, which means that reserved words are not allowed.
Directory names and URI path segments have no notion of Dart identifiers, so linking the shorthand to *only* Dart identifiers seems unnecessarily restrictive, and may disallow using the syntax for a URIs like:
```
import package.for.dart/banana; // contains `for`.
import mypkg/src/new/library; // contains `new`
```
Dart reserved words are small common words that *can* occur as directory names, or as parts of dotted names (if anyone used dotted names).

It's *easy* to allow reserved words in unquoted URIs.
Rather than using `<dottedIdentifierList>`, use:
```dart
<dottedUriWords> ::= (<uriWord> '.')* <uriWord> 
<uriWord> ::= <identifier> | <RESERVED_WORD>
```
It does nothing except allow reserved words too. Grammatically it should be no harder to work with than `<dottedIdentifierList>`, and it allows users to not worry about whether their file paths contain Dart reserved words, which have no reason to be special in that context.

Further, if we add more reserved words in the future, we will not introduce compile-time errors if someone used that word as a directory name.

(Why only disallow reserved words, and not built-in identifiers? If it's because there is no reason to disallow built-in identifiers, then there is equally no reason to disallow reserved words.)

We should consider that URIs can be followed by `if` in conditional imports:
```dart
import foo/bar.
  if (condition) foo/qux;
```
A mistaken trailing `.`, like after `bar` above, would include the `if` in the URI, then complain about the `(`.
That won't happen if we don't allow reserved words. It also won't happen if we disallow internal whitespace (#3983), and it's not significantly different from `import foo/bar. hide banana;`, which we do allow.
I don't think it's a reason to not allow reserved words.

### PR DESCRIPTION
**Pull request by munificent**
Fix #3983.
Fix #3984.

cc @dart-lang/language-team

### CODE REVIEW

**Review by lrhn — APPROVED** [review]
LGTM.
Nits and comments only, core functionality is sound. 
One suggestion: Don't allow `packagePath` after `part` or `part of` - I should probably file an issue.

**Comment by lrhn** [comment]
Maybe add at least one example using conditional imports:
```dart
import smarts/basic
  if (dart.library.io) smarts/native
  if (dart.library.js_interop) smarts/js;
```

File: working/unquoted-imports/feature-specification.md [hunk 1]
```diff
@@ -281,7 +315,37 @@ import strange /* comment */    .   but
...
+the one in the [analyzer package][], which is freely available), the reality is
+that users often cobble together simple scripts using regex to do this kind of
+parsing, or they need to write these tools in a language other than Dart. In
+those cases, if the package path happens to contain whitespace or comments, the
+tool will likely silently fail to recognize the package path.
+
+[analyzer package]: https://pub.dev/packages/analyzer
+
+Also, we find no compelling *use* for whitespace and comments inside package
+paths. To that end, this proposal makes it an error. All of the tokens in the
```
[the remark this answers is not in the document]
Comment (lrhn) [reply working/unquoted-imports/feature-specification.md:335]: Good point.

It's somewhat like inverted pyramid writing. Start by saying what this document is (it's the design document for a Dart language feature that will allow shorter syntax for common import URIs), then the conclusion and consequences. Then the actual specification. Then support it with further arguments and discussions. Then support it with extra data, graphs, etc., maybe in appendices. The goal is to give readers clear places to *get off* the reading train, when they have all they need, and so that they don't miss anything essential by doing so. If you don't care about specification for import URIs, you can stop after the summary. If you just want to know the big lines, you can stop after the first step. If you just want the spec, you can stop after step 2. If you want the *why*, you can keep reading.

(Academic writing tend to spend a lot of time on set-up and motivation. Design documents are about *decisions* first. The things we didn't choose are not going to be important to any of our users in the long run. It's nice to have the discussion of *why* things were chosen, so a later generation won't think "there was probably a good reason for this completely arbitrary decision" when it really was "it's as good as any, and we need one". But most readers just want the conclusion.)

File: working/unquoted-imports/feature-specification.md [hunk 2]
```diff
@@ -220,27 +220,30 @@ import flutter/material;
...
-flag directory or library names that aren't identifiers.
+This means you can't have a path segment that uses some combination of
+characters that isn't currently a single token in Dart, like `hyphen-separated`
+or `123LeadingDigits`. A path component must be an identifier (which may be a
+reserved word or built-in identifier, discussed below). Fortunately, our
+published guidance has *always* told users that [package names][name guideline]
+and [directories][directory guideline] should be valid Dart identifiers. Pub
+will complain if you try to publish a package whose name isn't a valid
+identifier. Likewise, the linter will flag directory or file names that aren't
+identifiers.
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/unquoted-imports/feature-specification.md:246]: Oh, interesting. Filed https://github.com/dart-lang/linter/issues/5071 to align with the proposal and tweaked the wording here.

File: working/unquoted-imports/feature-specification.md [hunk 3]
```diff
@@ -281,7 +315,37 @@ import strange /* comment */    .   but
...
+Also, we find no compelling *use* for whitespace and comments inside package
+paths. To that end, this proposal makes it an error. All of the tokens in the
+path must be directly adjacent with no whitespace, newlines, or comments between
+them. The previous import is an error. However, we still allow comments in or
+after the directives outside of the path. These are all valid:
+
+```dart
+import /* Weird but OK. */ some/path;
+export some/path; // Hi there.
+part some/path // Before the semicolon? Really?
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/unquoted-imports/feature-specification.md:343]: This should probably be filed as an issue so that we don't lose the discussion. :)

For now, I'll leave the proposal as-is, but I'm definitely keen to talk about this more. For what it's worth, I would ultimately like to support unquoted parts. It's true that we don't support relative unquoted references. But my guess is that users will end up preferring unquoted package references even to files in the same package versus quoted relative imports. At least, that's what I think *I'll* end up doing.

If that's the case, then unquoted part directives will make sense too.

File: working/unquoted-imports/feature-specification.md [hunk 4]
```diff
@@ -291,27 +355,33 @@ We add a new rule and hang it off the existing `uri` rule already used by import
 and export directives:
 
 ```
-uri                   ::= stringLiteral | packagePath
-packagePath           ::= packagePathSegment ( '/' packagePathSegment )*
-packagePathSegment    ::= dottedIdentifierList
-dottedIdentifierList  ::= identifier ('.' identifier)*
+uri               ::= stringLiteral | packagePath
+packagePath       ::= pathSegment ( '/' pathSegment )*
+pathSegment       ::= segmentComponent ( '.' segmentComponent )*
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/unquoted-imports/feature-specification.md:360]: Ooh, good catch. Went through and fixed everything.

File: working/unquoted-imports/feature-specification.md [hunk 5]
```diff
@@ -291,27 +355,33 @@ We add a new rule and hang it off the existing `uri` rule already used by import
...
+segmentComponent  ::= identifier
+                    | ⟨RESERVED_WORD⟩
+                    | ⟨BUILT_IN_IDENTIFIER⟩
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/unquoted-imports/feature-specification.md:363]: Done. (Went with the latter 'cause it's flatter.)

File: working/unquoted-imports/feature-specification.md [hunk 6]
```diff
@@ -291,27 +355,33 @@ We add a new rule and hang it off the existing `uri` rule already used by import
...
+                    | ⟨OTHER_IDENTIFIER⟩
 ```
 
-An import or export can continue to use a `stringLiteral` for the quoted form
-(which is what they will do for relative imports). But they can also use a
-`packagePath`, which is a slash-separated series of segments, each of which is a
-series of dot-separated identifiers. *(The `dottedIdentifierList` rule is
-already in the grammar and is shown here for clarity.)*
+It is a compile-time error if any whitespace, newlines, or comments occur
+between any of the `segmentComponent`, `/`, or `.` tokens in a `packagePath`.
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/unquoted-imports/feature-specification.md:368]: If I was implementing the parser, I would simply parse the grammar ignoring comments or whitespace between the tokens like you do in most places. Then I'd go back and report errors if there were comments or whitespace between any of the terminals.

So the proposal is pretty much specified the same way.

I don't think it matters too much one way or the other. It might matter a lot of we make semicolons optional, but I think we can cross that bridge when we come to it. I'll leave it as is and if the parser folks ask questions, I can tweak it.

File: working/unquoted-imports/feature-specification.md [hunk 7]
```diff
@@ -291,27 +355,33 @@ We add a new rule and hang it off the existing `uri` rule already used by import
...
-directive followed by an unquoted identifier like `part of;` or `part
-of.some/other.thing;` versus a `part of` directive like `part of thing;` or
-`part of 'uri.dart';` It must lookahead past the `of` identifier to see if the
-next token is `;`, `.`, `/`, or another identifier.*
+*There are two directives for working with part files, `part` and `part of`.
+This means that when the parser sees `part of`, it doesn't immediately know if
+it is looking at a `part` directive followed by an unquoted identifier like
+`part of;` or `part of.some/other.thing;` versus a `part of` directive like
+`part of thing;` or `part of 'uri.dart';` It must lookahead past the `of`
+identifier to see if the next token is `;`, `.`, `/`, or another identifier.*
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/unquoted-imports/feature-specification.md:384]: > If we allow that, then `part of deferred as f;` would be ambiguous, if seen by itself.

So you're saying this could be either:

* A `part of` directive for a file named `deferred` which should be locally prefixed as `f`.
* A `part` directive for a file named `of` which is deferred loaded and available under prefix `f`.

Is that right? If so, I don't think it's ambiguous since the latter isn't valid. We only allowed `deferred` on imports, not `part`, right? Does enhanced parts change that?

File: working/unquoted-imports/feature-specification.md [hunk 8]
```diff
@@ -354,14 +421,14 @@ a string literal containing that string. The process is:
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/unquoted-imports/feature-specification.md:421]: Good point, done.

File: working/unquoted-imports/feature-specification.md [hunk 9]
```diff
@@ -354,14 +421,14 @@ a string literal containing that string. The process is:
...
+        Otherwise, it's the last identifier after the last `.`. So in `foo`,
+        *path* is `foo`. In `foo.bar.baz`, it's `baz`.*
 
     3.  The URI is "package:*name*/*path*.dart". *So `import test;` desugars to
-        `import "package:test/test.dart";`, and `import server.api;` desugars
-        to `import "package:server.api/api.dart";`.*
+        `import "package:test/test.dart";`, and `import server.api;` desugars to
+        `import "package:server.api/api.dart";`.*
 
 5.  Else:
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/unquoted-imports/feature-specification.md:434]: Good catch! Fixed.

File: working/unquoted-imports/feature-specification.md [hunk 10]
```diff
@@ -463,7 +530,7 @@ this proposal's semantics. In other words, `part of foo.bar;` is part of the
 library at `package:foo/bar.dart`, not part of the library with name `foo.bar`.
```
[the remark this answers is not in the document]
Comment (munificent) [reply working/unquoted-imports/feature-specification.md:530]: Want to file an issue to discuss? I don't have a strong opinion, but whatever opinions we have should be captured somewhere they won't disappear.

File: working/unquoted-imports/feature-specification.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (munificent) [reply working/unquoted-imports/feature-specification.md:335]: I have many feelings about inverted pyramid and narrative order. :)

It totally makes sense to put the conclusions up front and let people get off the train early when they have the information they need.

At the same time, my experience is that in persuasive writing, it can be more effective to delay the conclusion until later in the narrative. When you state it up front, if the reader has a knee-jerk negative reaction to it, they will resist the rest of the prose and poke holes at it. If you lead them gently to the conclusion along a path of things they agree with, when they reach the end, it often sinks in more effectively.

The challenge is that language proposals are doing a mixture of informing and persuading, so figuring out the right order is hard.

But in this case, I think Kallen has a good point. Added a couple of sentences up front. I didn't add anything to the reserved words section since the previous section already states that reserved words and built-in identifiers are allowed.

File: working/unquoted-imports/feature-specification.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (lrhn) [reply working/unquoted-imports/feature-specification.md:345]: > preferring unquoted package references even to files in the same package versus quoted relative imports

I find it unlikely that most people will prefers `part mypkg/src/subfeature/featurepart;` to `part "featurepart.dart";` in `mypkg/src/subfeature.feature.dart`.

We don't even have a way to omit the package name, and parts will always at least be below `src/` too, and almost always in the same directory as the library file.

It also only works for libraries inside `lib/`. If you have part files in `test/`, `bin/`, `tool/` or (now) `hook/`, you can't use unquoted package URIs for those.

But we'll see :Grin:. (I expect users will ask for ways to not quote relative paths, and we *can* give them that, as `part ./featurepart;`. I still think we should use `:` to separate package name from path, because that makes it harder to typo a relative path into a package path. But the most likely mistake is to do `import foo;` meaning `import "./foo.dart";`, which doesn't have any separator, so the separator probably won't make a big difference.)

File: working/unquoted-imports/feature-specification.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (lrhn) [reply working/unquoted-imports/feature-specification.md:386]: That was what I thought when I wrote it, but you can't have deferred parts (well, yet! ... but likely ever) so that's not really ambiguous.

So what I should have said is: If we allow *something* after `part <uri>` in the future, and that *something* can be a single identifier or dotted identifier, then <code>part of *something*;</code> is ambiguous.

And if we allow things after both `part <uri>` and `part of <uri>`, then if any *thing* after `part <uri>` has the form `validUnquotedUri otherThing` where *otherThing` is valid after `part of <uri>`,  then `part of *thing*` is ambiguous.

A lot of "if"s.

The things I have actually considered so far are, during "parts with modifiers" discussions:
* `as id` modifiers on `part of` and `part` to contain the identifiers they introduce.
* `sealed` modifier on `part of` to just not inherit any imports from the parent file.

But, as Erik just pointed out to me, *until* we have a concrete ambiguity, we shouldn't be saying what an incorrect program means. That's a job for recovery and error messaging. We should say what allowed programs mean, and we should not allow programs that we don't want the user to write.

As currently specified, a valid program can have `part id.id/id.id/id;` with no spaces between the `id`s and the `.` and `/`s.
If it doesn't follow that format, it's not a valid Dart program, and we don't *have* to say how it's wrong, just that it is wrong.

That does mean that `part of;` is valid and means the same as `part "package:of/of.dart";`.
This is not parsing recovery if we allow the syntax and then give a "file/package not found" error. (It's incredibly unlikely that `package:of/of.dart` is a part file, and the current file is inside `package:foo`'s `lib` directory, which is the only way it can not be an error.)

It's still surprising. It may be error-prone, if someone writes `part of;` and forgets to write a URI. Maybe they removed the `library foo;` because we tell them to, and then change `part of foo;` to just `part of;`. (It's obvious which file it's part of, right?)

If we want to prevent that, we can:
* Not allow unquited URIs in `part` directives, or
* Not consider a single identifier `of` as an unquoted URI in a `part` directive. (Basically the same as saying that `of` is a contextually reserved word, and *contextually* reserved words take precedence over everything else if they occur where we reserved them.)

Or we *can* do nothing, and have our tools recognize and give a tailored error message if someone does `part of;` and `package:of/of.dart` fails to exist or be a valid part of the current file. ("You wrote..., you probably meant to write ...".)

File: working/unquoted-imports/feature-specification.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (lrhn) [reply working/unquoted-imports/feature-specification.md:532]: Did: https://github.com/dart-lang/language/issues/4038

[the remark this answers is not in the document]
Comment (munificent) [reply working/unquoted-imports/feature-specification.md:532]: Oops, didn't see that. Thanks!
